File: Level-2/70.java
import java.util.*;

class Solution {
    
    public int[][] arr;
    
    public int bfs(int n, int start) {
        int cnt = 1;
        boolean[] visit = new boolean[n + 1];
        Queue<Integer> q = new LinkedList<>();
        
        q.offer(start);
        while(!q.isEmpty()) {
            int point = q.poll();
            visit[point] = true;
            for(int i = 1; i <= n; i++) {
                if(visit[i]) continue;
                if(arr[point][i] == 1) {
                    q.offer(i);
                    cnt++;
                }
            }
        }
        
        // cnt - (n - cnt)
        return Math.abs(2 * cnt - n);
    }
    
    public int solution(int n, int[][] wires) {
        int answer = n;
        arr = new int[n + 1][n + 1];
        
        for(int i = 0; i < wires.length; i++) {
            arr[wires[i][0]][wires[i][1]] = 1;
            arr[wires[i][1]][wires[i][0]] = 1;
        }
        
        for(int i = 0; i < wires.length; i++) {
            int a = wires[i][0];
            int b = wires[i][1];
            
            // 선 끊고
            arr[a][b] = 0;
            arr[b][a] = 0;
            
            answer = Math.min(answer, bfs(n, a));
            
            // 다시 복구
            arr[a][b] = 1;
            arr[b][a] = 1;
        }
        
        return answer;
    }
}

File: Level-2/70.js
function solution(n, wires) {
    let answer = Number.MAX_SAFE_INTEGER;
    // 트리만들기
    let tree = Array.from(Array(n + 1), () => []);
    wires.map((element) => {
        let [a, b] = element;
        
        tree[a].push(b);
        tree[b].push(a);
    });
    
    // 제외숫자는 제외하고 탐색 (root노드, 예외노드)
    const searchTree = (root, except) => {
        let count = 0;
        let visit = [];
        let queue = [root];
        visit[root] = true;
        while(queue.length) {
            let index = queue.pop();
            tree[index].forEach((element) => {
                if(element !== except && !visit[element]) {
                    visit[element] = true;
                    queue.push(element);
                }
            });
            count++;
        }
        return count;
    };
    
    // 최솟값을 찾음
    wires.forEach((element) => {
        let [a, b] = element;
        answer = Math.min(answer, Math.abs(searchTree(a, b) - searchTree(b, a)));
    });
    
    return answer;
}
